Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -385,14 +385,42 @@ extension HTTPClientResponseDelegate {
     public func didReceiveError(task: HTTPClient.Task<Response>, _: Error) {}
 }
 
-internal extension URL {
+extension URL {
+    var percentEncodedPath: String {
+        if self.path.isEmpty {
+            return "/"
+        }
+        return self.path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? self.path
+    }
+
+    var pathHasTrailingSlash: Bool {
+        let url = self.absoluteString
+
+        var pathEndIndex = url.index(before: url.endIndex)
+        if let queryIndex = url.firstIndex(of: "?") {
+            pathEndIndex = url.index(before: queryIndex)
+        } else if let fragmentIndex = url.firstIndex(of: "#") {
+            pathEndIndex = url.index(before: fragmentIndex)
+        }
+
+        return url[pathEndIndex] == "/"
```

Comment:
Writing this against the absolute string seems to be making life quite hard: why not write it against `self.path`?

Reply:
`self.path` removes trailing slash